File: src/MatchingAlgorithm/ElectricCalculator.cs
namespace MatchingAlgorithm;

public static class TransformerCalculator
{
    /// <summary>
    /// Calculate voltage of primary side while retaining current and voltage limits
    /// </summary>
    /// <param name="expectedPower">Load power.</param>
    /// <param name="voltageMax">RMS voltage of primary side.</param>
    /// <param name="currentMax">RMS current of primary side.</param>
    /// <param name="nominalResistance">Resistance of primary side.</param>
    /// <param name="turnRatio">Turn ratio.</param>
    /// <param name="impedance">Impedance of system (secondary side).</param>
    public static double Voltage(
        double expectedPower,
        double voltageMax,
        double currentMax,
        double nominalResistance,
        double turnRatio,
        double impedance)
    {
        var resistanceMinimal = expectedPower / (currentMax * currentMax) / (turnRatio * turnRatio);
        var nominalSecondaryResistance = nominalResistance / (turnRatio * turnRatio);

        var voltage = voltageMax * Math.Sqrt(impedance / nominalSecondaryResistance);

        if (impedance < resistanceMinimal)
            voltage = currentMax / Math.Sqrt(8 / (Math.PI * Math.PI)) * impedance * turnRatio * turnRatio;

        return voltage > voltageMax ? voltageMax : voltage;
    }


    /// <summary>
    /// Calculate load power.
    /// </summary>
    /// <param name="resistance">Resistance of the system on the secondary side.</param>
    /// <param name="impedance">Impedance of the system on the secondary side.</param>
    /// <param name="voltage">RMS voltage on primary side.</param>
    /// <param name="turnRatio">Turn ratio.</param>
    public static double Power(double resistance, double impedance, double voltage, double turnRatio)
    {
        var current = Current(impedance, voltage, turnRatio);
        var power = current * voltage * resistance / impedance;
        power /= 1.11;
        return power;
    }


    /// <summary>
    /// Calculate current on primary side.
    /// </summary>
    /// <param name="impedance">Impedance of the system on the secondary side.</param>
    /// <param name="voltage">RMS voltage on primary side.</param>
    /// <param name="turnRatio">Turn ratio.</param>
    public static double Current(double impedance, double voltage, double turnRatio)
    {
        return voltage / (impedance * turnRatio * turnRatio) / 1.11;
    }
}

File: src/MatchingAlgorithm/ElectricCalculator.cpp
#include "ElectricCalculator.h"

#include <cmath>
#include <numbers>

double ElectricCalculator::CalculateVoltageLimit(const double power, const double maxVoltage, const double maxCurrent,
												 const double nominalResistance, const double turnRatio, const double impedance)
{
	const auto resistanceMinimal = power / (maxCurrent * maxCurrent) / (turnRatio * turnRatio);
	const auto nominalSecondaryResistance = nominalResistance / (turnRatio * turnRatio);

	auto voltage = maxVoltage * sqrt(impedance / nominalSecondaryResistance);

	if(impedance < resistanceMinimal)
		voltage = maxCurrent / sqrt(8 / (std::numbers::pi * std::numbers::pi)) * impedance * turnRatio * turnRatio;

	return voltage > maxVoltage ? maxVoltage : voltage;
}

double ElectricCalculator::CalculatePower(const double resistance, const double impedance, const double voltage, const double turnRatio)
{
	const auto current = CalculateCurrent(impedance, voltage, turnRatio);
	auto power = current * voltage * resistance / impedance;
	power /= 1.11;
	return power;
}

double ElectricCalculator::CalculateCurrent(const double impedance, const double voltage, const double turnRatio)
{
	return voltage / (impedance * turnRatio * turnRatio) / 1.11;
}
